src/communication/rpc_impl.rs

```diff
@@ -0,0 +1,569 @@
+}
+
+impl InMemoryPendingRequestStore {
+    #[cfg(test)]
+    fn contains(&self, reqid: &UUID) -> bool {
+        self.pending_requests
+            .lock()
+            .map_or(false, |pending_requests| {
+                pending_requests.contains_key(reqid)
+            })
+    }
+}
+
+impl PendingRequestStore for InMemoryPendingRequestStore {
+    fn try_add(&self, reqid: UUID) -> Result<ResponseFuture, ServiceInvocationError> {
+        if let Ok(mut pending_requests) = self.pending_requests.lock() {
+            let response_future_state = Arc::new(Mutex::new(ResponseFutureState::default()));
+            pending_requests.insert(reqid, response_future_state.clone());
+            Ok(ResponseFuture {
+                state: response_future_state,
+            })
+        } else {
+            Err(ServiceInvocationError::RpcError(UStatus::fail_with_code(
+                UCode::INTERNAL,
+                "failed to add response handler",
+            )))
+        }
+    }
+    fn complete_response_future(&self, reqid: &UUID, response_message: UMessage) {
+        if let Ok(mut pending_requests) = self.pending_requests.lock() {
+            if let Some(response_future_state) = pending_requests.remove(reqid) {
+                if let Ok(mut state) = response_future_state.lock() {
+                    state.response_message = Some(response_message);
+                    if let Some(waker) = state.waker.take() {
+                        waker.wake();
+                    }
+                }
+            }
+        }
+    }
+    fn remove(&self, reqid: &UUID) -> Option<Arc<Mutex<ResponseFutureState>>> {
+        if let Ok(mut pending_requests) = self.pending_requests.lock() {
+            pending_requests.remove(reqid)
+        } else {
+            None
+        }
+    }
+}
+
+struct ResponseListener {
+    pending_request_store: Arc<dyn PendingRequestStore>,
+}
+
+impl ResponseListener {
+    fn try_add_pending_request(
+        &self,
+        reqid: UUID,
+    ) -> Result<ResponseFuture, ServiceInvocationError> {
+        self.pending_request_store.try_add(reqid)
+    }
+    fn remove_pending_request(&self, reqid: &UUID) -> Option<Arc<Mutex<ResponseFutureState>>> {
+        self.pending_request_store.remove(reqid)
+    }
+}
+
+#[async_trait]
+impl UListener for ResponseListener {
+    async fn on_receive(&self, msg: UMessage) {
+        if msg
+            .attributes
+            .as_ref()
+            .map_or(false, |attribs| attribs.is_response())
+        {
+            if let Some(reqid) = msg
+                .attributes
+                .as_ref()
+                .and_then(|attribs| attribs.reqid.clone().into_option())
+            {
+                self.pending_request_store
+                    .complete_response_future(&reqid, msg)
+            }
+        }
+    }
+}
+
+/// An ['RpcClient'] which keeps all information about pending requests in memory.
+pub struct InMemoryRpcClient {
```

Comment:
In large scale (cloud) environments you could think of keeping the state in a shared persistent store to provide for horizontal scalability. Not sure, though, if this will ever be part of the up-rust crate. But FMPOV it still makes sense to make the implementation strategy transparent so that users know what they can expect.